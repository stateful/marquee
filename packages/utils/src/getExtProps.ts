import os from "os";
import fs from 'fs';
import path from 'path';
import vscode from "vscode";

const EXTENSION_ID = "stateful.marquee";
const TELEMETRY_CONFIG_ID = "telemetry";
const TELEMETRY_CONFIG_ENABLED_ID = "enableTelemetry";

export function getMarqueeVersion (context: vscode.ExtensionContext) {
  const extensionPath = path.join(context.extensionPath, "package.json");
  const packageFile = fs.readFileSync(extensionPath, "utf-8");
  const packageJson = JSON.parse(packageFile);
  return packageJson.version;
}

export default function getExtProps(context: vscode.ExtensionContext): { [key: string]: string } {
  const extProps = Object.create(null);

  const config = vscode.workspace.getConfiguration(TELEMETRY_CONFIG_ID);
  if (!config.get<boolean>(TELEMETRY_CONFIG_ENABLED_ID, true)) {
    return extProps;
  }

  extProps["os"] = os.platform();
  extProps["platformversion"] = (os.release() || "").replace(
    /^(\d+)(\.\d+)?(\.\d+)?(.*)/,
    "$1$2$3"
  );
  extProps["extname"] = EXTENSION_ID;
  extProps["extversion"] = getMarqueeVersion(context);
  if (vscode && vscode.env) {
    extProps["vscodemachineid"] = vscode.env.machineId;
    extProps["vscodesessionid"] = vscode.env.sessionId;
    extProps["vscodeversion"] = vscode.version;

    switch (vscode.env.uiKind) {
      case vscode.UIKind.Web:
        extProps["uikind"] = "web";
        break;
      case vscode.UIKind.Desktop:
        extProps["uikind"] = "desktop";
        break;
      default:
        extProps["uikind"] = "unknown";
    }
  }
  return extProps;
}
