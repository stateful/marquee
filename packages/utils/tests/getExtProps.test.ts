import getExtProps from '../src/getExtProps';

jest.mock('../src/constants', () => ({
  pkg: { version: '1.2.3' }
}));
jest.mock('os', () => ({
  platform: () => 'some platform',
  release: () => 'some release'
}));
jest.mock('vscode', () => ({
  workspace: { getConfiguration: jest.fn().mockReturnValue(new Map([
    ['enableTelemetry', true]
  ])) },
  env: {
    machineId: 'machineId',
    sessionId: 'sessionId',
    uiKind: 'uiKind'
  },
  version: '3.2.1',
  UIKind: { Web: 'uiKind' }
}));

test('should get proper extension props', () => {
  expect(getExtProps()).toMatchSnapshot();
});
