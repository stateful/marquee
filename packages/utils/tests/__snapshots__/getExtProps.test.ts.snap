// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should get proper extension props 1`] = `
Object {
  "extname": "stateful.marquee",
  "extversion": "1.2.3",
  "os": "some platform",
  "platformversion": "some release",
  "uikind": "web",
  "vscodemachineid": "machineId",
  "vscodesessionid": "sessionId",
  "vscodeversion": "3.2.1",
}
`;
